File: Algorithm/DynamicProgramming/atCoder/Knapsack DP/B19/B19.php
<!-- 以下は、**PHP 8.2.8** を使用して、**Knapsack 2（価値軸DP）** を解くための実装です。

## ✅ 特徴

* **関数分割済み**
* **引数と戻り値の型指定**
* **処理時間・メモリ制約（10秒・1024MiB）を考慮**
* `int` 型は PHP 内部で 64bit まで扱える（W = 1e9 に対応可能）

## ✅ 使用関数一覧と説明

| 関数名               | 説明                | パラメータ・返却値                                                        |
| ----------------- | ----------------- | ---------------------------------------------------------------- |
| `parseInput()`    | 標準入力からデータを取得して構造化 | 戻り値: `array{int, int, array<array{int, int}>}`                   |
| `solveKnapsack()` | DP配列で最大の価値を計算     | 引数: `int $N, int $W, array<array{int,int}> $items`<br>返り値: `int` |
| `main()`          | 全体統括（入出力）         | 戻り値: `void`                                                      |

---

## ✅ PHPコード（PHP 8.2.8）

<?php

/**
 * 標準入力から値を読み込み、品物情報を構造化して返す
 *
 * @return array{int, int, array<array{int, int}>} [N, W, items]
 */
function parseInput(): array {
    $lines = explode("\n", trim(file_get_contents("php://stdin")));
    [$N, $W] = array_map('intval', explode(' ', array_shift($lines)));
    $items = [];

    foreach ($lines as $line) {
        [$w, $v] = array_map('intval', explode(' ', $line));
        $items[] = [$w, $v]; // [重さ, 価値]
    }

    return [$N, $W, $items];
}

/**
 * ナップザック問題を価値軸DPで解き、最大の価値を返す
 *
 * @param int $N 品物数
 * @param int $W ナップサック容量
 * @param array<array{int, int}> $items 品物の [重さ, 価値] の配列
 * @return int 容量W以下で得られる最大の価値
 */
function solveKnapsack(int $N, int $W, array $items): int {
    $maxValue = 0;
    foreach ($items as [$w, $v]) {
        $maxValue += $v;
    }

    $INF = PHP_INT_MAX;
    $dp = array_fill(0, $maxValue + 1, $INF);
    $dp[0] = 0;

    foreach ($items as [$w, $v]) {
        for ($val = $maxValue; $val >= $v; --$val) {
            if ($dp[$val - $v] !== $INF && $dp[$val - $v] + $w <= $W) {
                $dp[$val] = min($dp[$val], $dp[$val - $v] + $w);
            }
        }
    }

    // dp[v] <= W を満たす最大の v を探す
    for ($v = $maxValue; $v >= 0; --$v) {
        if ($dp[$v] <= $W) {
            return $v;
        }
    }

    return 0;
}

/**
 * メイン関数：標準入力を読み取り、解を出力
 *
 * @return void
 */
function main(): void {
    [$N, $W, $items] = parseInput();
    $result = solveKnapsack($N, $W, $items);
    echo $result . PHP_EOL;
}

main();

// ## ✅ 入力例（input.txt）

// ```
// 4 7
// 3 13
// 3 17
// 5 29
// 1 10
// ```

// ### 実行方法

// ```bash
// php main.php < input.txt
// ```

// ## ✅ 計算量・メモリ見積もり

// | 指標     | 値                                    |
// | ------ | ------------------------------------ |
// | 時間計算量  | `O(N × sum_v)` ≈ 100 × 100000 = 10^7 |
// | メモリ使用量 | 約 400 KB（dp配列：100001 × 8バイト）         |
// | 実行時間   | 数百ミリ秒〜1秒以下（十分高速）                     |
// | メモリ制限  | 1024 MiB 制限内に収まる                     |

// ## ✅ 補足

// * PHPは `array` 型が柔軟なので、型アノテーションをコメントで補っています。
// * より厳格な型検査をしたい場合は [Psalm](https://psalm.dev/) のような静的解析ツールも有効です。

File: Algorithm/DynamicProgramming/atCoder/Knapsack DP/B19/B19.py
# 以下は、**Python (CPython 3.11.4)** を使用して、**Knapsack 2（価値軸DP）** を解く実装です。関数を分割し、**型アノテーション付き**で、**処理時間・メモリの効率にも配慮**しています。

# ---

# ## ✅ 構成と特徴

# | 関数名                | 役割          | 返却値                                      |
# | ------------------ | ----------- | ---------------------------------------- |
# | `parse_input()`    | 入力パース処理     | `tuple[int, int, list[tuple[int, int]]]` |
# | `solve_knapsack()` | DPによる最大価値計算 | `int`（価値の最大値）                            |
# | `main()`           | 入出力処理       | `None`（標準出力）                             |

# ---

# ## ✅ Python 実装（関数分割 + コメント付き）

# ```python
import sys
from typing import List, Tuple

def parse_input() -> Tuple[int, int, List[Tuple[int, int]]]:
    """
    入力をパースして返す関数

    Returns:
        N (int): 品物数
        W (int): ナップサックの容量
        items (List[Tuple[int, int]]): 各品物の (重さ, 価値) のリスト
    """
    lines = sys.stdin.read().strip().split('\n')
    N, W = map(int, lines[0].split())
    items: List[Tuple[int, int]] = [
        (int(w), int(v)) for w, v in (line.split() for line in lines[1:])
    ]
    return N, W, items

def solve_knapsack(N: int, W: int, items: List[Tuple[int, int]]) -> int:
    """
    ナップザック問題を価値軸DPで解く

    Args:
        N (int): 品物数
        W (int): ナップサック容量
        items (List[Tuple[int, int]]): 各品物の (重さ, 価値)

    Returns:
        int: 容量 W 以下で達成可能な最大の価値
    """
    max_value = sum(v for _, v in items)
    INF = float('inf')
    dp: List[float] = [INF] * (max_value + 1)
    dp[0] = 0  # 価値0に必要な重さは0

    for w, v in items:
        for val in range(max_value, v - 1, -1):
            if dp[val - v] + w <= W:
                dp[val] = min(dp[val], dp[val - v] + w)

    # dp[v] <= W を満たす最大の v を返す
    for val in range(max_value, -1, -1):
        if dp[val] <= W:
            return val
    return 0


def main() -> None:
    """
    メイン関数：標準入力から読み取り、最大価値を出力

    Returns:
        None
    """
    N, W, items = parse_input()
    result = solve_knapsack(N, W, items)
    print(result)


if __name__ == "__main__":
    main()
# ```

# ---

# ## ✅ 実行例

# ```bash
# python3 main.py < input.txt
# ```

# ---

# ## ✅ 計算量・メモリ見積もり

# | 指標     | 内容                                     |
# | ------ | -------------------------------------- |
# | 時間計算量  | O(N × max\_value) = 100 × 100000 ≒ 1e7 |
# | メモリ使用量 | dp\[100001] × 8バイト ≒ 約 800 KB          |
# | 実行時間制限 | 10 秒以内（通常 0.1～0.5秒で終了）                 |
# | メモリ制限  | 1024 MiB（このコードは余裕あり）                   |

# ---

# ## ✅ 処理ステップ図（概要）

# ```text
# 標準入力 → parse_input()
#               ↓
#           N, W, items
#               ↓
#        solve_knapsack()  ← DPテーブル生成
#               ↓
#            最大価値を算出
#               ↓
#            print(result)
# ```

# ---

# ## ✅ 拡張案（必要なら）

# * 品物選択の **経路復元（どの品物を選んだか）**
# * 重さ軸DP（Wが小さい場合のみ有効）
# * メモリ最適化（1次元DP）
